extractor/filesystem/embeddedfs/vdi/vdi.go

```diff
@@ -0,0 +1,931 @@
+		if err != nil {
+			return fmt.Errorf("failed to close raw file %s: %w", e.tmpRawPath, err)
+		}
+		if err := os.Remove(e.tmpRawPath); err != nil {
+			return fmt.Errorf("failed to remove temporary raw file %s: %w", e.tmpRawPath, err)
+		}
+	}
+	if err := os.RemoveAll(e.tmpDir); err != nil {
+		return fmt.Errorf("failed to remove temporary directory %s: %w", e.tmpDir, err)
+	}
+	return nil
+}
+
+// ntfsDirFS wraps scalibrfs.DirFS to include reference counting and cleanup
+type ntfsDirFS struct {
+	fs         scalibrfs.FS
+	file       *os.File
+	tmpDir     string
+	tmpRawPath string
+	refCount   *int32
+	refMu      *sync.Mutex
+}
+
+func (n *ntfsDirFS) Open(name string) (fs.File, error) {
+	return n.fs.Open(name)
+}
+
+func (n *ntfsDirFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	if name == "/" || name == "" {
+		name = "."
+	}
+	return n.fs.ReadDir(name)
+}
+
+func (n *ntfsDirFS) Stat(name string) (fs.FileInfo, error) {
+	if name == "/" || name == "" || name == "." {
+		// Return synthetic FileInfo for root directory
+		return &fileInfo{
+			name:    name,
+			isDir:   true,
+			modTime: time.Now(),
+		}, nil
+	}
+	return n.fs.Stat(name)
+}
+
+func (n *ntfsDirFS) Close() error {
+	n.refMu.Lock()
+	defer n.refMu.Unlock()
+	if n.file == nil {
+		return nil // Already closed
+	}
+	*n.refCount--
+	if *n.refCount == 0 {
+		err := n.file.Close()
+		n.file = nil // Prevent double close
+		if err != nil {
+			return fmt.Errorf("failed to close raw file %s: %w", n.tmpRawPath, err)
+		}
+		if err := os.Remove(n.tmpRawPath); err != nil {
+			return fmt.Errorf("failed to remove temporary raw file %s: %w", n.tmpRawPath, err)
+		}
+	}
+	if err := os.RemoveAll(n.tmpDir); err != nil {
+		return fmt.Errorf("failed to remove temporary directory %s: %w", n.tmpDir, err)
+	}
+	return nil
+}
+
+// fileInfo is a simple implementation of fs.FileInfo for the root directory
+type fileInfo struct {
+	name    string
+	isDir   bool
+	modTime time.Time
+}
+
+func (fi *fileInfo) Name() string {
+	return fi.name
+}
+
+func (fi *fileInfo) Size() int64 {
+	return 0
+}
+
+func (fi *fileInfo) Mode() fs.FileMode {
+	if fi.isDir {
+		return fs.ModeDir | 0755
+	}
+	return 0644
+}
+
+func (fi *fileInfo) ModTime() time.Time {
+	return fi.modTime
+}
+
+func (fi *fileInfo) IsDir() bool {
+	return fi.isDir
+}
+
+func (fi *fileInfo) Sys() any {
+	return nil
+}
+
+// VDI conversion functions
+
+func convertVDIToRaw(inPath string, outPath string) error {
+	f, err := os.Open(inPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Read header
+	var hdr header
+	if err := binary.Read(f, binary.LittleEndian, &hdr); err != nil {
+		return err
+	}
+
+	// Sanity check: VDI signature should be 0xBEDA107F
+	if hdr.Signature != Signature {
+		return errors.New("not a valid VDI file (bad signature)")
+	}
+
+	// Open output
+	out, err := os.Create(outPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	switch hdr.ImageType {
+	case 2: // fixed/static
```

Comment:
Which  header type does the test VDI use?
It would be great if we had a test file for both cases.

Reply:
Hmm sure.

I tend to include the most complicate header type inside the test images. From personal experience, fixed/static is not as complicated as sparse / dynamic in the same format. If I have to rank file formats based on decreasing complexity, then it's gonna be like this : Qemu > Vmware  > Microsoft > VirtualBox